src/data/polynomial/module.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+
+import ring_theory.finiteness
+
+/-!
+# Polynomial module
+
+In this file, we define the polynomial module for an `R`-module `M`, i.e. the `R[X]`-module `M[X]`.
+
+This is defined as an type alias `polynomial_module R M := ℕ →₀ M`, since there might be different
+module structures on `ℕ →₀ M` of interest. See the docstring of ``polynomial_module` for details.
+
+-/
+
+universes u v
+
+open polynomial
+open_locale polynomial big_operators
+
+variables (R M : Type*) [comm_ring R] [add_comm_group M] [module R M] (I : ideal R)
+
+include R
+
+/--
+The `R[X]`-module `M[X]` for an `R`-module `M`.
+
+We require all the module instances `module S (polynomial_module R M)` to factor through `R` except
+`module R[X] (polynomial_module R M)`.
+In this constraint, we have the following instances for example :
+- `R` acts on `polynomial_module R R[X]`.
+- `R[X]` acts on `polynomial_module R R[X]` as `R[Y]` acting on `R[X][Y]`
+- `R[X]` acts on `polynomial_module R[X] R[X]`.
+- `R[X]` acts on `polynomial_module R[X] R[X]` as `R[X]` acting on `R[X][Y]`
+- `R[X][X]` acts on `polynomial_module R[X] R[X]` as `R[X][Y]` acting on itself.
+
+This is also the reason why `R` is included in the alias, or else there will be two different
+instances of `module R[X] (polynomial_module R[X])`.
```

Comment:
I think there should be a remark in this docstring about `polynomial R R` vs `R[X]`. Are they defeq as types? As `R`-module? As `R[X]`-module? I think the answer to all three questions is no. But I'm not sure.